src/app.test.js

```diff
@@ -1,15 +1,52 @@
-// import { createApp } from "./app";
-// import userevent from '@testing-library/user-event'
-
-test("renders the todo app", () => {
-  //   // setup
-  //   const root = ...
-  //   const container = createApp(root)
-  //   // act
-  //   const input = queryByPlaceholderText('Please enter todo here')
-  //   userevent.type(input, 'new task')
-  //   const addButton = queryByText('Add')
-  //   userevent('click', addButton)
-  //   // assert
-  expect(true).tobee(true);
+import { screen, getByText } from "@testing-library/dom";
+import userEvent from "@testing-library/user-event";
+import createApp from "./app";
+// import userEvent from "@testing-library/user-event";
+
+test("renders empty todo list when starts", () => {
+  // setup
+  const root = document.body;
+  createApp(root);
+  const todoList = screen.getByTestId("todo-list");
+  // act
+  // assert
+  expect(todoList).toBeEmptyDOMElement();
+});
+
+test("shows new item in todo list when added", () => {
+  // setup
+  const input = screen.getByPlaceholderText("Please enter todo here");
+  const addButton = screen.getByTestId("add-button");
+  const inputText = "Take wife for a walk!";
+
+  // act
+  userEvent.type(input, inputText);
+  userEvent.click(addButton);
+
+  // assert
+  const todoList = screen.getByTestId("todo-list");
+  const todoTextElement = getByText(todoList, "Take wife for a walk!");
+  expect(todoTextElement).toHaveTextContent(inputText);
+});
+
+test("should delete an item from todo List", () => {
+  // setup
+  const input = screen.getByPlaceholderText("Please enter todo here");
+  const addButton = screen.getByTestId("add-button");
+  const inputText = "Todo to be deleted!";
+
+  // act
+  userEvent.type(input, inputText);
+  userEvent.click(addButton);
+  const todoList = screen.getByTestId("todo-list");
+  const todoTextElement = getByText(todoList, "Todo to be deleted!");
+  expect(todoList.childElementCount).toBe(2);
+
+  const todoElement = todoTextElement.parentElement;
+  const todoDeleteButton = getByText(todoElement, "Delete");
+  userEvent.click(todoDeleteButton);
+
+  // assert
+  expect(todoList).not.toContainElement(todoElement);
+  expect(todoList.childElementCount).toBe(1);
```

Comment:
again, this is tightly coupled with the implementation: it's not possible to change (or refactor) the production code without breaking this assertion:

```js
// these would fail in case we reorganised the DOM e.g. by adding an intermediate element between the todo list element and the todo item element
expect(todoList).not.toContainElement(todoElement);
expect(todoList.childElementCount).toBe(1);
```

in general assertions against the application should be created from the user's perspective, so instead (as the user would do) it's enough to find out that the todo we've added in the _setup_ phase cannot be found in the document anymore:

```js
expect(getByText(todoList, "Take wife for a walk!")).not.toBeInTheDocument();
```

allowing us to freely reorganise the code and the DOM structure as much as possible